src/indexer.c

```diff
@@ -126,59 +130,113 @@ int CountPredicateType(QueryPredicateList *queries, PredicateType type) {
-        if (RedisModule_DictSize(leaf) == 0) {
-            RedisModule_FreeDict(NULL, leaf);
-            RedisModule_DictDel(labelsIndex, key, NULL);
-        }
+        RedisModule_DictSet(ts_leaf, key, NULL);
+    } else if (op & Indexer_Remove) {
+        labelsIndexRemoveTsKey(leaf, key, ts_key, _labelsIndex);
     }
 }
 
-void IndexOperation(RedisModuleCtx *ctx,
-                    INDEXER_OPERATION_T op,
-                    RedisModuleString *ts_key,
-                    Label *labels,
-                    size_t labels_count) {
+void IndexMetric(RedisModuleString *ts_key, Label *labels, size_t labels_count) {
     const char *key_string, *value_string;
     for (int i = 0; i < labels_count; i++) {
         size_t _s;
         key_string = RedisModule_StringPtrLen(labels[i].key, &_s);
         value_string = RedisModule_StringPtrLen(labels[i].value, &_s);
         RedisModuleString *indexed_key_value =
-            RedisModule_CreateStringPrintf(ctx, KV_PREFIX, key_string, value_string);
-        RedisModuleString *indexed_key = RedisModule_CreateStringPrintf(ctx, K_PREFIX, key_string);
+            RedisModule_CreateStringPrintf(NULL, KV_PREFIX, key_string, value_string);
+        RedisModuleString *indexed_key = RedisModule_CreateStringPrintf(NULL, K_PREFIX, key_string);
 
-        indexUnderKey(op, indexed_key_value, ts_key);
-        indexUnderKey(op, indexed_key, ts_key);
+        labelIndexUnderKey(Indexer_Add, indexed_key_value, ts_key, labelsIndex, tsLabelIndex);
+        labelIndexUnderKey(Indexer_Add, indexed_key, ts_key, labelsIndex, tsLabelIndex);
 
-        RedisModule_FreeString(ctx, indexed_key_value);
-        RedisModule_FreeString(ctx, indexed_key);
+        RedisModule_FreeString(NULL, indexed_key_value);
+        RedisModule_FreeString(NULL, indexed_key);
     }
 }
 
-void IndexMetric(RedisModuleCtx *ctx,
-                 RedisModuleString *ts_key,
-                 Label *labels,
-                 size_t labels_count) {
-    IndexOperation(ctx, Indexer_Add, ts_key, labels, labels_count);
+// Removes the ts from the label index and from the inverse index, if exist.
+// del_key should be false if caller wants to avoid iterator invalidation.
+void RemoveIndexedMetric_generic(RedisModuleString *ts_key,
+                                 RedisModuleDict *_labelsIndex,
+                                 RedisModuleDict *_tsLabelIndex,
+                                 bool del_key) {
+    int nokey = 0;
+    RedisModuleDict *ts_leaf = RedisModule_DictGet(_tsLabelIndex, ts_key, &nokey);
+    if (nokey) { // series has no labels or already been removed from index
+        return;
+    }
+
+    RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(ts_leaf, "^", NULL, 0);
+    RedisModuleString *currentLabelKey;
+    while ((currentLabelKey = RedisModule_DictNext(NULL, iter, NULL)) != NULL) {
+        labelIndexUnderKey(Indexer_Remove, currentLabelKey, ts_key, _labelsIndex, _tsLabelIndex);
```

Comment:
the function overloading here is a bit confusing, it doesn't really need the `_tsLabelIndex` because it only removes from the label->key index, right?

Reply:
It removes also from label->key:
Watch the following lines:
    if (del_key) {
        RedisModule_DictDel(_tsLabelIndex, ts_key, NULL);
    }